sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java

```diff
@@ -24,78 +42,226 @@
+            if (table.getSchema().getName().equals("PUBLIC")) { // H2-specific assumption
                 tables.put(table.getName(), Arrays.asList(table.fields()));
             }
         }
     }
 
+    /*
+     * All executed SQL queries against a JOOQ connection are received here
+     */
     @Override
     public MockResult[] execute(final MockExecuteContext ctx) throws SQLException {
-        final MockResult[] mock = new MockResult[1];
-        // The execute context contains SQL string(s), bind values, and other meta-data
-        final String sql = ctx.sql();
-
-        if (sql.toUpperCase().startsWith("SELECT")) {
-            mock[0] = executeSelectStar(sql);
-        } else {
-            // Exceptions are propagated through the JDBC and jOOQ APIs
-            throw new SQLException("Statement not supported: " + sql);
+        final String[] batchSql = ctx.batchSQL();
+        final MockResult[] mock = new MockResult[batchSql.length];
+        try {
+            dDlogAPI.transactionStart();
+            for (int i = 0; i < batchSql.length; i++) {
+                mock[i] = executeOne(batchSql[i]);
+            }
+            dDlogAPI.transactionCommit();
+        } catch (final DDlogException e) {
+            throw new RuntimeException(e);
         }
         return mock;
     }
 
-    private MockResult executeSelectStar(final String sql) throws SQLException {
-        final String[] s = sql.toUpperCase().split(" ");
-        if (!(s.length == 4 && s[1].equals("*") && s[2].equals("FROM"))) {
-            throw new SQLException("Statement not supported: " + sql);
+    private MockResult executeOne(final String sql) throws SQLException {
+        final Statement statement = parser.createStatement(sql, options);
+        final MockResult result = queryVisitor.process(statement, sql);
+        if (result == null) {
+            throw new SQLException("Could not execute SQL statement " + sql);
         }
-        final String tableName = s[3];
-        final List<Field<?>> fields = tables.get(tableName);
-        if (fields == null) {
-            throw new SQLException("Unknown table: " + tableName);
+        return result;
+    }
+
+    /*
+     * Visits an SQL query and converts into a JOOQ MockResult type.
+     */
+    private class QueryVisitor extends AstVisitor<MockResult, String> {
+        @Override
+        protected MockResult visitQuerySpecification(final QuerySpecification node, String sql) {
+            // The checks below encode assumption A1 (see javadoc for the DDlogJooqProvider class)
+            final Select select = node.getSelect();
+            if (!(select.getSelectItems().size() == 1 && select.getSelectItems().get(0) instanceof AllColumns)) {
+                throw new RuntimeException("Statement not supported: " + sql);
+            }
+            assert node.getFrom().isPresent() && node.getFrom().get() instanceof com.facebook.presto.sql.tree.Table;
```

Comment:
this is a bit inconsistent, sometimes you throw, sometimes you assert.

Reply:
Done!